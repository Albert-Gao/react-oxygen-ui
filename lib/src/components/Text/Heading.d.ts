import React from 'react';
import { IText } from './Text';
export declare const Heading: React.FC<IText>;
