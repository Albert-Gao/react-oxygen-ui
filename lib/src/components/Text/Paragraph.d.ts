import React from 'react';
import { IText } from '../';
export declare const Paragraph: React.FC<IText>;
