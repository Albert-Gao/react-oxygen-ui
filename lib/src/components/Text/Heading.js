"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const styles_1 = require("../../styles");
const Text_1 = require("./Text");
exports.Heading = ({ children, size, ...rest }) => {
    let tagName = 'h1';
    if (size === styles_1.DEFAULT_FONT_SIZE.heroHeading) {
        tagName = 'h1';
    }
    else if (size === styles_1.DEFAULT_FONT_SIZE.mainHeading) {
        tagName = 'h2';
    }
    else if (size === styles_1.DEFAULT_FONT_SIZE.pageHeading) {
        tagName = 'h3';
    }
    else if (size === styles_1.DEFAULT_FONT_SIZE.sectionHeading) {
        tagName = 'h4';
    }
    else if (size === styles_1.DEFAULT_FONT_SIZE.bodyHeading) {
        tagName = 'h5';
    }
    else {
        tagName = 'h6';
    }
    return (<Text_1.Text size={size} as={tagName} fontWeight="bold" {...rest}>
      {children}
    </Text_1.Text>);
};
