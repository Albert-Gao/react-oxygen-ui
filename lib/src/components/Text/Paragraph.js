"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const __1 = require("../");
// It uses <Text /> under the hook, the main differences are it will
// add max-width to enforce a best practice in terms of readability.
// how many words a line.
exports.Paragraph = ({ size, noDefaultLineHeight = false, ...rest }) => (<__1.Text as="p" {...__1.getFontSpecStyle(noDefaultLineHeight, true, size)} {...rest}/>);
