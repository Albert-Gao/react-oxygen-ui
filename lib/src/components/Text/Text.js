"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const styled_system_1 = require("styled-system");
const fontSpecs_1 = require("../../styles/theme/constants/fontSpecs");
const utils_1 = require("../../utils/utils");
const BaseBox_1 = require("../Box/BaseBox");
function addStyleToResult(tempStyle, result, key) {
    if (utils_1.is.notExist(result[key])) {
        result[key] = [];
    }
    // @ts-ignore
    // tslint:disable-next-line: no-unsafe-any
    result[key].push(tempStyle[key]);
}
exports.getFontSpecStyle = (noDefaultLineHeight, isMaxWidthIncluded = false, size) => {
    if (utils_1.is.notExist(size)) {
        return {};
    }
    let result = {};
    if (Array.isArray(size)) {
        if (size.length === 1) {
            result = exports.getFontSpecStyle(noDefaultLineHeight, isMaxWidthIncluded, size[0]);
        }
        else {
            const temp = [];
            size.forEach(s => {
                const style = exports.getFontSpecStyle(noDefaultLineHeight, isMaxWidthIncluded, s);
                temp.push(style);
            });
            temp.forEach(tempStyle => {
                addStyleToResult(tempStyle, result, 'fontSize');
                if (!noDefaultLineHeight) {
                    addStyleToResult(tempStyle, result, 'lineHeight');
                }
                if (isMaxWidthIncluded) {
                    addStyleToResult(tempStyle, result, 'maxWidth');
                }
            });
        }
    }
    else {
        const fontSpecInTheme = fontSpecs_1.fontSpecs[size];
        const isInTheme = typeof fontSpecInTheme === 'object';
        if (isInTheme) {
            result = {
                fontSize: styled_system_1.px(fontSpecInTheme.size),
                lineHeight: fontSpecInTheme.lineHeight,
            };
            if (isMaxWidthIncluded) {
                result.maxWidth = styled_system_1.px(fontSpecInTheme.maxWidth);
            }
        }
        else {
            result = {
                fontSize: styled_system_1.px(size),
            };
        }
    }
    if (result === {}) {
        result = { fontSize: size };
    }
    return result;
};
exports.Text = ({ size, noDefaultLineHeight = false, takeRef, ...rest }) => (<BaseBox_1.BaseBox as="div" {...exports.getFontSpecStyle(noDefaultLineHeight, false, size)} {...utils_1.getRefProp(takeRef)} {...rest}/>);
