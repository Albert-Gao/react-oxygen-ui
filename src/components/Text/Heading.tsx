import React from 'react';
import { DEFAULT_FONT_SIZE } from '../../styles';
import { IText, Text } from './Text';

export const Heading: React.FC<IText> = ({
  children,
  size,
  ...rest
}) => {
  let tagName: keyof JSX.IntrinsicElements = 'h1';

  if (size === DEFAULT_FONT_SIZE.heroHeading) {
    tagName = 'h1';
  } else if (size === DEFAULT_FONT_SIZE.mainHeading) {
    tagName = 'h2';
  } else if (size === DEFAULT_FONT_SIZE.pageHeading) {
    tagName = 'h3';
  } else if (size === DEFAULT_FONT_SIZE.sectionHeading) {
    tagName = 'h4';
  } else if (size === DEFAULT_FONT_SIZE.bodyHeading) {
    tagName = 'h5';
  } else {
    tagName = 'h6';
  }

  return (
    <Text size={size} as={tagName} fontWeight="bold" {...rest}>
      {children}
    </Text>
  );
};
