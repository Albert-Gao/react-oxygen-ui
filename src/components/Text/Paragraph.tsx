import React from 'react';
import { getFontSpecStyle, IText, Text } from '../';

// It uses <Text /> under the hook, the main differences are it will
// add max-width to enforce a best practice in terms of readability.
// how many words a line.
export const Paragraph: React.FC<IText> = ({
  size,
  noDefaultLineHeight = false,
  ...rest
}) => (
  <Text
    as="p"
    {...getFontSpecStyle(noDefaultLineHeight, true, size)}
    {...rest}
  />
);
