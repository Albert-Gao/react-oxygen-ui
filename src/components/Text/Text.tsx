import React from 'react';
import { px } from 'styled-system';
import { ThemeType } from '../../styles';
import {
  fontSpecs,
  IFontSpecType,
} from '../../styles/theme/constants/fontSpecs';
import { getRefProp, is } from '../../utils/utils';
import { BaseBox } from '../Box/BaseBox';
import { IBaseBoxWithTakeRef, Responsive } from '../Box/BaseBox.type';

export interface IText extends IBaseBoxWithTakeRef {
  noDefaultLineHeight?: boolean;

  // so, it works like a handy shortcut.
  // if you set a size which can be indexed in theme.fontSizes
  // then fontSize, lineHeight and maxWidth(for Paragraph only) will all be set
  // if it's some random value, like '7px', it will be passed to final result directly
  // but won't include lineHeight and maxWidth, b/c we can't locate theme in theme.fontSpecs
  // it can be set as an array for responsive style which is the same behavior from styled-system
  size?: Responsive;
}

interface IFontSpecStyle {
  fontSize?: Responsive;
  lineHeight?: Responsive;
  maxWidth?: Responsive;
}

function addStyleToResult(
  tempStyle: IFontSpecStyle,
  result: IFontSpecStyle,
  key: keyof IFontSpecStyle,
) {
  if (is.notExist(result[key])) {
    result[key] = [];
  }

  // @ts-ignore
  // tslint:disable-next-line: no-unsafe-any
  result[key].push(tempStyle[key]);
}

export const getFontSpecStyle = (
  noDefaultLineHeight: boolean,
  isMaxWidthIncluded: boolean = false,
  size?: Responsive,
): IFontSpecStyle => {
  if (is.notExist(size)) {
    return {};
  }

  let result: IFontSpecStyle = {};

  if (Array.isArray(size)) {
    if (size.length === 1) {
      result = getFontSpecStyle(
        noDefaultLineHeight,
        isMaxWidthIncluded,
        size[0],
      );
    } else {
      const temp: IFontSpecStyle[] = [];

      size.forEach(s => {
        const style = getFontSpecStyle(
          noDefaultLineHeight,
          isMaxWidthIncluded,
          s,
        );
        temp.push(style);
      });

      temp.forEach(tempStyle => {
        addStyleToResult(tempStyle, result, 'fontSize');

        if (!noDefaultLineHeight) {
          addStyleToResult(tempStyle, result, 'lineHeight');
        }

        if (isMaxWidthIncluded) {
          addStyleToResult(tempStyle, result, 'maxWidth');
        }
      });
    }
  } else {
    const fontSpecInTheme = fontSpecs[
      size as keyof ThemeType['fontSpecs']
    ] as IFontSpecType;
    const isInTheme = typeof fontSpecInTheme === 'object';

    if (isInTheme) {
      result = {
        fontSize: px(fontSpecInTheme.size),
        lineHeight: fontSpecInTheme.lineHeight,
      };

      if (isMaxWidthIncluded) {
        result.maxWidth = px(fontSpecInTheme.maxWidth);
      }
    } else {
      result = {
        fontSize: px(size),
      };
    }
  }

  if (result === {}) {
    result = { fontSize: size };
  }

  return result;
};

export const Text: React.FC<IText> = ({
  size,
  noDefaultLineHeight = false,
  takeRef,
  ...rest
}) => (
  <BaseBox
    as="div"
    {...getFontSpecStyle(noDefaultLineHeight, false, size)}
    {...getRefProp(takeRef)}
    {...rest}
  />
);
