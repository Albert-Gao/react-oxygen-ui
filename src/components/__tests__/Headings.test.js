import 'jest-styled-components';
import 'jest-dom/extend-expect';
import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { Heading } from '../Text/Heading';
import {
  ThemeProvider,
  fontSpecs,
  DEFAULT_FONT_SIZE,
  defaultTheme,
} from '../../styles';
import { getMediaQueryStr } from './Text.test';

afterEach(cleanup);

describe('Test <Headings />: ', () => {
  test('size props === heroHeading works', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Heading size={DEFAULT_FONT_SIZE.heroHeading} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild.tagName).toEqual('H1');
    expect(firstChild).toHaveStyleRule(
      'font-size',
      fontSpecs[DEFAULT_FONT_SIZE.heroHeading].size,
    );
  });

  test('size props === bodyHeading works', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Heading size={DEFAULT_FONT_SIZE.bodyHeading} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild.tagName).toEqual('H5');
    expect(firstChild).toHaveStyleRule(
      'font-size',
      fontSpecs[DEFAULT_FONT_SIZE.bodyHeading].size,
    );
  });

  test('size props === [body, sectionHeading] works', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Heading size={[DEFAULT_FONT_SIZE.body, DEFAULT_FONT_SIZE.sectionHeading]} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild.tagName).toEqual('H6');

    expect(firstChild).toHaveStyleRule(
      'font-size',
      fontSpecs[DEFAULT_FONT_SIZE.body].size,
    );
    expect(firstChild).toHaveStyleRule(
      'line-height',
      fontSpecs[DEFAULT_FONT_SIZE.body].lineHeight.toString(),
    );
    expect(firstChild).not.toHaveStyleRule();

    expect(firstChild).toHaveStyleRule(
      'font-size',
      fontSpecs[DEFAULT_FONT_SIZE.sectionHeading].size,
      {
        media: getMediaQueryStr(defaultTheme.breakpoints[0]),
      },
    );

    expect(firstChild).toHaveStyleRule(
      'line-height',
      fontSpecs[DEFAULT_FONT_SIZE.sectionHeading].lineHeight.toString(),
      {
        media: getMediaQueryStr(defaultTheme.breakpoints[0]),
      },
    );

    expect(firstChild).not.toHaveStyleRule('max-width');
  });

  test('size props === [7px, sectionHeading] works', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Heading size={['7px', DEFAULT_FONT_SIZE.sectionHeading]} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild.tagName).toEqual('H6');

    expect(firstChild).toHaveStyleRule('font-size', '7px');
    expect(firstChild).not.toHaveStyleRule('line-height');
    expect(firstChild).not.toHaveStyleRule('max-width');

    expect(firstChild).toHaveStyleRule(
      'font-size',
      fontSpecs[DEFAULT_FONT_SIZE.sectionHeading].size,
      {
        media: getMediaQueryStr(defaultTheme.breakpoints[0]),
      },
    );

    expect(firstChild).toHaveStyleRule(
      'line-height',
      fontSpecs[DEFAULT_FONT_SIZE.sectionHeading].lineHeight.toString(),
      {
        media: getMediaQueryStr(defaultTheme.breakpoints[0]),
      },
    );

    expect(firstChild).not.toHaveStyleRule('max-width');
  });
});
