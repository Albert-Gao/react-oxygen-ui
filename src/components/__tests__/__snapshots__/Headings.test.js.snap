// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Headings />:  size props === [7px, sectionHeading] works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 7px;
  font-weight: 700;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

@media screen and (min-width:768px) {
  .c0 {
    font-size: 24px;
  }
}

@media screen and (min-width:768px) {
  .c0 {
    line-height: 1.61803398875;
  }
}

<h6
  class="c0"
  font-size="7px,24px"
  font-weight="bold"
/>
`;

exports[`Test <Headings />:  size props === [body, sectionHeading] works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.61803398875;
  font-weight: 700;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

@media screen and (min-width:768px) {
  .c0 {
    font-size: 24px;
  }
}

@media screen and (min-width:768px) {
  .c0 {
    line-height: 1.61803398875;
  }
}

<h6
  class="c0"
  font-size="14px,24px"
  font-weight="bold"
/>
`;

exports[`Test <Headings />:  size props === bodyHeading works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 16px;
  line-height: 1.61803398875;
  font-weight: 700;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

<h5
  class="c0"
  font-size="16px"
  font-weight="bold"
/>
`;

exports[`Test <Headings />:  size props === heroHeading works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 64px;
  line-height: 1.3;
  font-weight: 700;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

<h1
  class="c0"
  font-size="64px"
  font-weight="bold"
/>
`;

exports[`Test <Text />:  size props === [7px, sectionHeading] works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 7px;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

@media screen and (min-width:768px) {
  .c0 {
    font-size: 24px;
  }
}

@media screen and (min-width:768px) {
  .c0 {
    line-height: 1.61803398875;
  }
}

<div
  class="c0"
  font-size="7px,24px"
/>
`;

exports[`Test <Text />:  size props === [body, sectionHeading] with noDefaultLineHeight works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 14px;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

@media screen and (min-width:768px) {
  .c0 {
    font-size: 24px;
  }
}

<div
  class="c0"
  font-size="14px,24px"
/>
`;

exports[`Test <Text />:  size props === [body, sectionHeading] works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.61803398875;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

@media screen and (min-width:768px) {
  .c0 {
    font-size: 24px;
  }
}

@media screen and (min-width:768px) {
  .c0 {
    line-height: 1.61803398875;
  }
}

<div
  class="c0"
  font-size="14px,24px"
/>
`;

exports[`Test <Text />:  size props === [heroHeading] works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.61803398875;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

<div
  class="c0"
  font-size="14px"
/>
`;

exports[`Test <Text />:  size props === 7px works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 7px;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

<div
  class="c0"
  font-size="7px"
/>
`;

exports[`Test <Text />:  size props === heroHeading works 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 14px;
  line-height: 1.61803398875;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
}

<div
  class="c0"
  font-size="14px"
/>
`;
